doc/manual/syntax.md

```diff
@@ -0,0 +1,313 @@
+
+Examples:
+```
+> 1 == 1
+true
+
+> 5 == 5.0
+true
+
+> "Hello" == "Hello"
+true
+
+> "Hello" != "World"
+true
+
+> 5 == "Hello"
+false
+
+> true == "true"
+false
+```
+
+
+## Composite values
+
+### List
+A list is a sequence of values. They are delimited by `[` and `]`, and elements are separated with `,`.
+
+Examples:
+```
+[1, 2, 3]
+["Hello", "World"]
+[1, true, "true"]
+[]
+```
+
+Lists can be concatenated with the operator `@`:
+```
+> [1] @ [2, 3]
+[ 1, 2, 3 ]
+```
+
+### Record
+Records are key-value storage, or in Nickel terms, field-value storage. They are delimited by `{` and `}`, and elements are separated by `,`.
+Field-value elements are noted as `field = value`.
+The fields are strings, but can be noted as identifiers if they respect identifiers syntax (internally, they will be stored as strings). Values can be of any type.
+Elements inside a record are unordered.
+Two records can be _merged_ together using the operator `&`. The reader can find more information about merging in the relevant documentation.
+
+Examples:
+```
+{}
+{a = 3}
+{my_id_n5 = "my id number 5", my_id_n4 = }
+{"5" = 5, six = 6}
+```
+
+Accessing a record field can be done using the `.` operator :
+```
+> { a = 1, b = 5 }.a
+1
+
+> { a = 1 }.b
+error: Missing field
+
+> { "1" = "one" }."1"
+"one"
+```
+
+It is possible to write records of records via the *pieces syntax*, where we separate fields by dots:
+```
+> { a = { b = 1 } }
+{ a = { b = 1 } }
+ 
+> { a.b = 1 }
+{ a = { b = 1 } }
+
+> { a.b = 1, a.c = 2, b = 3}
+{ a = { b = 1, c = 2 }, b = 3 }
+```
+
+## Statements
+
+### If-Then-Else
+This statement allows conditional branching in your code. You can use it as `if <bool expr> then <expr> else <expr>`. Both expressions in the `then` and `else` branches are lazily evaluated, and must have the same type.
+
+Examples:
+```
+> if true then "TRUE :)" else "false :("
+"TRUE :)"
+
+> if false then "Not this one" else "This one"
+"This one"
+
+> if "forty-two" == 42 then "equal?" else "unequal"
+"unequal"
+```
+
+### Let-In
+Let-in allows the binding of an expression. It is used as `let <ident> = <expr> in <expr>`.
+
+Examples:
+```
+> let r = { a = "a", b = "b" } in r.a
+"a"
+
+> let inner = { inside = true } in let outer = { outside = inner.inside } in outer.outside
+true
+
+> let a = 1 in let b = 2 in a + b
+3
+```
+
+## Functions
+A function is declared using the `fun` keyword, then arguments separated by spaces, and finally an arrow `=>` to add the body of the function.
+To call a function, just add the arguments after it separated with spaces.
+Functions in Nickel are curried, meaning that a function taking multiple arguments is actually a function that takes a single argument and returns a function taking the rest of the arguments, until it is applied.
+
+Example
+```
+> (fun a b => a + b) 1 2
+3
+
+let add = fun a b => a + b in add 1 2
+3
+
+> let add = fun a b => a + b in
+    let add1 = add 1 in
+        add1 2
+3
+```
+
```

Comment:
Should also talk about the curried operator syntax. For each infix operator `+`, `-`, `@`, etc. there is an associated function `(op)`, such as `(@)`, and `(+)` that turns it into a function. It can be (partially) applied and passed as a function:

```
>let plus_one = (+) 1 in
let times_two = (*) 2 in
plus_one (times_two 2)
5

>let flatten = lists.fold (@) [] in
flatten [[1,2], [3,4]]
[ 1, 2, 3, 4 ]
```

We can also introduce the reverse application operator `|>` here. `x |> f` is `f x`. It is left associative, such that `x |> f |> g` is `g (f  x)`. It is useful to chain operations on a value:

```
[1,2,3,4,5,6]
|> lists.map ((+) 1)
|> lists.filter ((<=) 5) // this is fun x =>  5 <= x, so elements GREATER than 5
|> lists.elem_at 0
// 5
```

Reply:
I didn't know they existed, this is so cool

Reply:
Yes, this is why this is a good thing you're writing this document :slightly_smiling_face: the knowledge won't be hidden in the code and our heads anymore. This syntax exists in other ML descendants, like Haskell and OCaml.